File: Strand/Screens/TrainingLoadCard.swift
import Foundation
import SwiftUI
import Charts
import StrandDesign
import StrandAnalytics
import WhoopStore

// MARK: - Training Load card (CTL / ATL / TSB)
//
// The first UI surface for the long-horizon training-load model (TrainingLoadEngine, added with the
// paired `ReadinessEngine.evaluateWithTrainingLoad`). It overlays chronic load (CTL, the 42-day
// fitness proxy) and acute load (ATL, the 7-day fatigue proxy); the gap between the two lines IS the
// TSB / "form" (CTL − ATL), surfaced as the headline number and a footer stat.
//
// Descriptive only: CTL/ATL/TSB never feed the Readiness level or any score, and the loads are NOOP's
// daily Effort/strain — NOT TRIMP. Long-horizon by nature, so the card models the full history rather
// than the Trends range window (14+ contiguous days are needed before anything is drawn).
//
// Isolated in its own file on purpose: TrendsView already sits near the iOS type-check budget, so this
// keeps its own inference cost out of that body.
struct TrainingLoadCard: View {
    let days: [DailyMetric]

    // yyyy-MM-dd → Date (en_US_POSIX, UTC) — same keying TrendsView uses so the x-axis matches.
    private static let dayParser: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    private struct Row: Identifiable {
        let date: Date
        let ctl: Double
        let atl: Double
        var id: Date { date }
    }

    /// One modelled point per day of the contiguous suffix the engine returned.
    private var rows: [Row] {
        result.points.compactMap { p in
            guard let d = Self.dayParser.date(from: p.day) else { return nil }
            return Row(date: d, ctl: p.chronicLoad, atl: p.acuteLoad)
        }
    }

    /// Model straight from the training-load engine — NOT the paired `evaluateWithTrainingLoad`, which
    /// would also run the full Readiness synthesis this card never uses. `DailyMetric.strain` is the load.
    private var result: TrainingLoadEngine.Result {
        let loads = days.map { TrainingLoadEngine.DailyLoad(day: $0.day, load: $0.strain) }
        return TrainingLoadEngine.evaluate(days: loads)
    }

    private static let established = TrainingLoadEngine.Configuration.standard.establishedDays
    private static let minimum = TrainingLoadEngine.Configuration.standard.minimumDays

    private func fmt(_ v: Double) -> String { String(format: "%.1f", v) }
    private func signed(_ v: Double) -> String { String(format: "%+.1f", v) }

    var body: some View {
        let tl = result
        if !tl.isAvailable {
            unavailableCard(contiguousDays: tl.contiguousDays)
        } else {
            let latest = tl.points.last
            ChartCard(
                title: "Training Load",
                subtitle: subtitle(for: tl),
                trailing: latest.map { signed($0.balance) },
                height: NoopMetrics.chartHeight,
                chart: {
                    VStack(alignment: .leading, spacing: NoopMetrics.space2) {
                        legend
                        chart
                    }
                },
                footer: {
                    ChartFooter([
                        ("CTL", latest.map { fmt($0.chronicLoad) } ?? "—"),
                        ("ATL", latest.map { fmt($0.acuteLoad) } ?? "—"),
                        ("Form", latest.map { signed($0.balance) } ?? "—"),
                        ("Days", "\(tl.contiguousDays)"),
                    ])
                }
            )
        }
    }

    // Two overlaid lines: CTL (fitness) and ATL (fatigue). The vertical gap between them is the form.
    private var chart: some View {
        // Floor at 1 (matching the Android `fold(1.0)` twin): an all-rest window of zero loads would
        // otherwise make the y-domain `0...0`, which Swift Charts renders as a degenerate/empty scale.
        let maxY = max(rows.map { max($0.ctl, $0.atl) }.max() ?? 1, 1)
        return Chart {
            ForEach(rows) { r in
                LineMark(x: .value("Day", r.date), y: .value("CTL", r.ctl),
                         series: .value("Series", "CTL"))
                    .foregroundStyle(StrandPalette.gold)
                    .interpolationMethod(.catmullRom)
            }
            ForEach(rows) { r in
                LineMark(x: .value("Day", r.date), y: .value("ATL", r.atl),
                         series: .value("Series", "ATL"))
                    .foregroundStyle(StrandPalette.strain100)
                    .interpolationMethod(.catmullRom)
            }
        }
        .chartYScale(domain: 0...(maxY * 1.08))
        .chartYAxis { AxisMarks(position: .leading) }
        .accessibilityLabel(Text("Training load: chronic vs acute"))
    }

    private var legend: some View {
        HStack(spacing: NoopMetrics.space2 * 2) {
            legendDot(color: StrandPalette.gold, label: "CTL · Fitness")
            legendDot(color: StrandPalette.strain100, label: "ATL · Fatigue")
            Spacer()
        }
    }

    private func legendDot(color: Color, label: LocalizedStringKey) -> some View {
        HStack(spacing: NoopMetrics.space2) {
            Circle().fill(color).frame(width: 8, height: 8)
            Text(label).font(StrandFont.footnote).foregroundStyle(StrandPalette.textTertiary)
        }
    }

    private func subtitle(for tl: TrainingLoadEngine.Result) -> String {
        switch tl.state {
        case .established:
            return String(localized: "42-day fitness vs 7-day fatigue")
        case .building:
            return String(localized: "Building — \(tl.contiguousDays) of \(Self.established) days")
        case .unavailable:
            return ""
        }
    }

    // Honest empty state: name exactly how many consecutive Effort days are still needed.
    private func unavailableCard(contiguousDays: Int) -> some View {
        ChartCard(
            title: "Training Load",
            subtitle: String(localized: "Chronic vs acute load"),
            chart: {
                VStack(spacing: NoopMetrics.space2) {
                    Text("Needs \(Self.minimum)+ consecutive days of Effort to begin. \(contiguousDays) so far.")
                        .font(StrandFont.subhead)
                        .foregroundStyle(StrandPalette.textTertiary)
                        .multilineTextAlignment(.center)
                }
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .center)
            },
            footer: { EmptyView() }
        )
    }
}

File: android/app/src/main/java/com/noop/ui/TrainingLoadCard.kt
package com.noop.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.material3.Text
import com.noop.R
import com.noop.analytics.TrainingLoadEngine
import com.noop.data.DailyMetric
import kotlin.math.max

/**
 * Training Load card (CTL / ATL / TSB) — the first UI surface for [TrainingLoadEngine], twin of the
 * Apple `TrainingLoadCard`. Overlays chronic load (CTL, 42-day fitness proxy) and acute load (ATL,
 * 7-day fatigue proxy) on ONE shared y-scale, so the vertical gap between the lines reads as the TSB /
 * "form" (CTL − ATL), surfaced as the headline number + a footer stat.
 *
 * Descriptive only: CTL/ATL/TSB never feed the Charge/Readiness score, and the load is NOOP's daily
 * Effort/strain — not TRIMP. Long-horizon by nature, so it models the full history (not the range
 * window) and shows an honest "needs N more days" state until 14+ contiguous days exist.
 */
@Composable
fun TrainingLoadCard(days: List<DailyMetric>, modifier: Modifier = Modifier) {
    // Model straight from the training-load engine — NOT the paired evaluateWithTrainingLoad, which
    // would also run the full Readiness synthesis this card never uses. DailyMetric.strain is the load.
    val loads = days.map { TrainingLoadEngine.DailyLoad(it.day, it.strain) }
    val result = TrainingLoadEngine.evaluate(loads)
    val latest = result.points.lastOrNull()

    val subtitle = when (result.state) {
        TrainingLoadEngine.State.ESTABLISHED -> stringResource(R.string.trends_tl_established)
        TrainingLoadEngine.State.BUILDING ->
            stringResource(R.string.trends_tl_building, result.contiguousDays, TrainingLoadEngine.standard.establishedDays)
        TrainingLoadEngine.State.UNAVAILABLE -> stringResource(R.string.trends_tl_unavailable)
    }

    NoopCard(modifier = modifier) {
        Column(verticalArrangement = Arrangement.spacedBy(Metrics.space12)) {
            Row(verticalAlignment = Alignment.Top) {
                Column(Modifier.weight(1f)) {
                    Overline(text = stringResource(R.string.trends_training_load))
                    Text(subtitle, style = NoopType.footnote, color = Palette.textTertiary)
                }
                if (latest != null) {
                    Text(signed(latest.balance), style = NoopType.bodyNumber, color = Palette.textPrimary)
                }
            }

            if (!result.isAvailable) {
                Text(
                    stringResource(
                        R.string.trends_tl_needs,
                        TrainingLoadEngine.standard.minimumDays,
                        result.contiguousDays,
                    ),
                    style = NoopType.body,
                    color = Palette.textTertiary,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth().padding(vertical = Metrics.space12),
                )
            } else {
                Row(horizontalArrangement = Arrangement.spacedBy(Metrics.space12)) {
                    LegendDot(Palette.gold, stringResource(R.string.trends_tl_legend_ctl))
                    LegendDot(Palette.metricAmber, stringResource(R.string.trends_tl_legend_atl))
                }
                TwoLineChart(result.points, Modifier.fillMaxWidth().height(Metrics.chartHeight))
                Row(Modifier.fillMaxWidth()) {
                    // CTL/ATL are universal training-science acronyms (like HRV/SpO2) — same in every language.
                    Stat("CTL", latest?.let { fmt(it.chronicLoad) } ?: "—")
                    Stat("ATL", latest?.let { fmt(it.acuteLoad) } ?: "—")
                    Stat(stringResource(R.string.trends_tl_form), latest?.let { signed(it.balance) } ?: "—")
                    Stat(stringResource(R.string.trends_days), "${result.contiguousDays}")
                }
            }
        }
    }
}

@Composable
private fun TwoLineChart(points: List<TrainingLoadEngine.Point>, modifier: Modifier) {
    val ctl = Palette.gold
    val atl = Palette.metricAmber
    // Shared y-scale over BOTH series so the gap between the lines is a faithful TSB.
    val maxY = points.fold(1.0) { acc, p -> max(acc, max(p.chronicLoad, p.acuteLoad)) }
    Canvas(modifier) {
        val n = points.size
        if (n < 2) return@Canvas
        val w = size.width
        val h = size.height
        fun path(value: (TrainingLoadEngine.Point) -> Double): Path {
            val p = Path()
            points.forEachIndexed { i, pt ->
                val x = w * i / (n - 1)
                val y = h - (value(pt) / maxY).toFloat() * h
                if (i == 0) p.moveTo(x, y) else p.lineTo(x, y)
            }
            return p
        }
        drawPath(path { it.chronicLoad }, color = ctl, style = Stroke(width = 3f))
        drawPath(path { it.acuteLoad }, color = atl, style = Stroke(width = 3f))
    }
}

@Composable
private fun LegendDot(color: Color, label: String) {
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Metrics.space4)) {
        Spacer(Modifier.size(8.dp).clip(CircleShape).background(color))
        Text(label, style = NoopType.footnote, color = Palette.textTertiary)
    }
}

@Composable
private fun androidx.compose.foundation.layout.RowScope.Stat(label: String, value: String) {
    Column(Modifier.weight(1f)) {
        Text(label, style = NoopType.footnote, color = Palette.textTertiary)
        Text(value, style = NoopType.captionNumber, color = Palette.textSecondary)
    }
}

private fun fmt(v: Double): String = String.format(java.util.Locale.US, "%.1f", v)
private fun signed(v: Double): String = String.format(java.util.Locale.US, "%+.1f", v)
